Apps/Playground/node_modules/react-native-babylon/android/src/main/cpp/BabylonNativeInterop.cpp

```diff
@@ -0,0 +1,104 @@
+#include <jni.h>
+
+#include <Babylon/JsRuntime.h>
+#include <Babylon/NativeWindow.h>
+#include <Babylon/NativeEngine.h>
+
+#include <arcana/threading/task_schedulers.h>
+
+#include <JavaScriptCore/JavaScript.h>
+
+#include <android/log.h>
+#include <android/native_window.h>
+#include <android/native_window_jni.h>
+
+#include <optional>
+#include <sstream>
+#include <unistd.h>
+
+namespace Babylon
+{
+    namespace
+    {
+        void log(const char *str)
+        {
+            __android_log_print(ANDROID_LOG_VERBOSE, "BabylonNative", "%s", str);
+        }
+    }
+
+    class Native final
+    {
+    public:
+        // This class must be constructed from the JavaScript thread
+        Native(JSGlobalContextRef jsContext, ANativeWindow* windowPtr)
+        {
+            m_env = std::make_shared<Napi::Env>(Napi::Attach(jsContext));
+            m_looper_scheduler = std::make_shared<looper_scheduler_t>(looper_scheduler_t::get_for_current_thread());
+
+            JsRuntime::DispatchFunctionT dispatchFunction{[weak_env = std::weak_ptr<Napi::Env>(m_env), weak_looper_scheduler = std::weak_ptr<looper_scheduler_t>(m_looper_scheduler)](std::function<void(Napi::Env)> func) {
+                auto looper_scheduler = weak_looper_scheduler.lock();
+                if (looper_scheduler)
+                {
+                    (*looper_scheduler)([weak_env, func = std::move(func)]()
+                    {
+                        auto env = weak_env.lock();
+                        if (env)
+                        {
+                            func(*env);
+                        }
+                    });
+                }
+            }};
+
+            JsRuntime::CreateForJavaScript(*m_env, dispatchFunction);
+
+            auto width = static_cast<size_t>(ANativeWindow_getWidth(windowPtr));
+            auto height = static_cast<size_t>(ANativeWindow_getHeight(windowPtr));
+
+            InitializeGraphics(windowPtr, width, height);
+            InitializeNativeEngine(*m_env);
+            NativeWindow::Initialize(*m_env, windowPtr, width, height);
+
+            // TODO: This shouldn't be necessary, but for some reason results in a significant increase in frame rate. Need to figure this out.
+            ReinitializeNativeEngine(*m_env, windowPtr, width, height);
+        }
+
+        ~Native()
+        {
+            DeinitializeGraphics();
+        }
+
+        void Refresh(ANativeWindow* windowPtr)
+        {
+            auto width = static_cast<size_t>(ANativeWindow_getWidth(windowPtr));
+            auto height = static_cast<size_t>(ANativeWindow_getHeight(windowPtr));
+            ReinitializeNativeEngine(*m_env, windowPtr, width, height);
+        }
+
+    private:
+        using looper_scheduler_t = arcana::looper_scheduler<sizeof(std::weak_ptr<Napi::Env>) + sizeof(std::function<void(Napi::Env)>)>;
+        std::shared_ptr<looper_scheduler_t> m_looper_scheduler;
+        std::shared_ptr<Napi::Env> m_env;
```

Comment:
I'm not sure either of these should be shared pointers. The `Env` almost certainly doesn't need to be; it can just be captured by copy in the `JsRuntime`'s dispatch function, which will then hold onto it forever. I'm less sure about the looper. Does it need to be owned by this? Or can it just be `std::move`ed into the dispatch function so that the `JsRuntime` can own it instead of just referencing it?

Reply:
Good point. I need to leave the looper_scheduler as a shared_ptr because looper_scheduler is not copyable, but the shared_ptr can be owned by the dispatch function.